ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/OracularClock.hs

```diff
@@ -0,0 +1,242 @@
+{-# LANGUAGE Rank2Types          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Intended for qualified import
+--
+-- > import Test.Util.OracularClock (OracularClock(..))
+-- > import qualified Test.Util.OracularClock as OracularClock
+module Test.Util.HardFork.OracularClock (
+    OracularClock(..)
+  , new
+  , onSlotChange
+  , withinEachSlot
+  , withinTheCurrentSlot
+  ) where
+
+import           Control.Monad (replicateM_, void)
+import           Data.Foldable (toList)
+import           Data.Time
+import           GHC.Stack
+
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Cardano.Slotting.Slot
+
+import qualified Ouroboros.Consensus.BlockchainTime as BTime
+import           Ouroboros.Consensus.Util.STM (onEachChange)
+import           Ouroboros.Consensus.Util.Time (nominalDelay)
+
+import           Test.Util.HardFork.Future (Future, futureSlotLengths,
+                     futureTimeToSlot)
+import           Test.Util.Slots (NumSlots (..))
+import           Test.Util.Stream
+
+-- | A clock that knows the future
+--
+-- This clock's closure contains a 'BTime.SystemTime', a 'Future', and a
+-- 'NumSlots'. As the system time advances, the clock maintains a notion of the
+-- current 'SlotNo'. Once all 'NumSlots' have passed, the clock is /exhausted/
+-- and most of its methods block indefinitely from then on.
+--
+-- Most notably, 'waitUntilDone' instead blocks until the the clock is
+-- exhausted; so the continuation of that call should expect other threads
+-- using the rest of this clock's methods to be stuck (and eg reap them).
+--
+-- Note: though the wallclock-slot correspondence depends on the ledger state,
+-- we have designed our ledgers so that all nodes necessarily use the same
```

Comment:
Not so much the _ledgers_ right? It's more than we (you) have set up all _tests_ so that no matter which chain is chosen, the future in terms of slot lengths and epoch sizes is always the same?

Reply:
The comment is about the fact that there is only one future, not multiple.

The ledgers are designed so that decisions about time are only determined by the content of the immutable prefix. The sentence also says "in the absence of a CP violation", and so all nodes share the same immutable prefix. Thus there is exactly one future. And that's _the_ future that the oracular clock correctly anticipates.

Does that clarify? Does it make sense? Let me know, and I'll try to refine the comment.

Reply:
But it's not in the design of the _ledger_. It's in the design of the transactions we _submit_ to the ledger. We make sure that no matter which chain is picked, no matter how forks are resolved, the _Future_ always looks the same.

Reply:
> But it's not in the design of the ledger.

I'm specifically thinking of the stability windows being defined on a per-ledger basis. That seems particularly relevant. Am I missing a distinction somewhere?

Reply:
Ah, I see, you're referring to the safe zones. But that's still not enough right. Outside of those safe zones the `Future` might look different, and I guess the tests _still_ need to make sure that, say, an A-B transition happens in the same place on every work, or else the oracular clock won't be able to perform it's miracles.

Reply:
We decided to revisit this later. Merging now.